### PR TITLE
Add local system searching/completing facility for contrib/ql.

### PR DESCRIPTION
**Pull request by teeghee**
Slime-quicklisp contributed package has lacked one important feature: searching and completing local systems.  The current pull request is for adding this facility in this package.  

The local project directory in quicklisp (usually this is `$HOME/quicklisp/local-projects/` contains system-index.txt file, in which the information of local asd files is stored.  In this new version of slime-quicklisp, it searches the asd files by reading system-index.txt and adds local system names into the completion collection.

### CODE REVIEW

File: contrib/swank-quicklisp.lisp [hunk 1]
```diff
@@ -14,4 +14,53 @@
...
+        (recur ins nil)))))
+
+(defslimefun list-quicklisp-local-systems
+    (&optional (local-system-dirs
+                (symbol-value
+                 (find-symbol "*LOCAL-PROJECT-DIRECTORIES*" "QL"))))
+  "Returns the Quicklisp local systems list."
+  (mapcan #'extract-system-names
+          (mapcan #'read-system-index-file local-system-dirs)))
+
```
Comment (luismbo) [inline contrib/swank-quicklisp.lisp]: Could we use `ql:list-local-systems` instead?

File: contrib/slime-quicklisp.el [hunk 2]
```diff
@@ -21,12 +21,13 @@
...
 
-
 (defun slime-read-quicklisp-system-name (&optional prompt default-value)
   "Read a Quick system name from the minibuffer, prompting with PROMPT."
   (let* ((completion-ignore-case nil)
          (prompt (or prompt "Quicklisp system"))
-         (quicklisp-system-names (slime-eval `(swank:list-quicklisp-systems)))
+         (quicklisp-system-names
+          (append (slime-eval `(swank:list-quicklisp-local-systems))
+                  (slime-eval `(swank:list-quicklisp-systems))))
```
Comment (luismbo) [inline contrib/slime-quicklisp.el]: Maybe just have `swank:list-quicklisp-systems` include local systems too, possibly subject to an optional argument.

File: contrib/swank-quicklisp.lisp (hunk 1 above)
Comment (teeghee) [reply contrib/swank-quicklisp.lisp]: Oh my, I was stupidly reinventing the wheel.  I will change this to make use of `ql:list-local-systems` function.

**Comment by teeghee** [comment]
I have no idea why the Travis CI has failed at ccl.  Could someone hint me?

**Comment by luismbo** [comment]
@teeghee the tests are not super robust, retrying the CCL run fixed the failure 🤷‍♂️

File: contrib/swank-quicklisp.lisp [hunk 3]
```diff
@@ -6,12 +6,16 @@
...
   (if (member :quicklisp *features*)
       (let ((ql-dist-name (find-symbol "NAME" "QL-DIST"))
-            (ql-system-list (find-symbol "SYSTEM-LIST" "QL")))
-        (mapcar ql-dist-name (funcall ql-system-list)))
+            (ql-system-list (find-symbol "SYSTEM-LIST" "QL"))
+            (ql-local-system-list
+             (find-symbol "LIST-LOCAL-SYSTEMS" "QL")))
+        (append (when with-local-systems
+                  (funcall ql-local-system-list))
+                (mapcar ql-dist-name (funcall ql-system-list))))
```
Comment (luismbo) [inline contrib/swank-quicklisp.lisp]: Should we sort them by name or something?

Comment (teeghee) [reply contrib/swank-quicklisp.lisp]: I don't think it is necessary: completing-read seems to sort the items (in this case the appended list in the above code block) automatically.

**Pull request closed without merging**